Abstract record retrieval behind an interface

smarterclayton opened the pull request:
Moves retrieval of records behind an interface as discussed in #116.
Still has some rough edges (PTRRecords, TTLs may not be correct,
single record lookups are possibly wrong).

Does this look roughly like what we discussed?

smarterclayton commented:
@thockin FYI this would remove the need to do the sync, at the cost of using a Service cache.  However, it's likely masters should have a service cache anyway.

miekg commented:
I think it looks very nice already. Doing a code review here on github sucks, so I think I'll send you a pr on  this pr after I tested and played with it (if a pr on a pr actually works)

smarterclayton commented:
If I'm not mistaken, the reverse lookup is wrong as I have it - I think I need to convey that to Record via a Boolean or a new method (reverse bool)?


> On Mar 7, 2015, at 2:50 AM, Miek Gieben <notifications@github.com> wrote:
> 
> I think it looks very nice already. Doing a code review here on github sucks, so I think I'll send you a pr on this pr after I tested and played with it (if a pr on a pr actually works)
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented:
My records lookup was definitely wrong - updated the pull with a new method `ReverseRecord(name string) (*msg.Service, error)`

The kube example is under a build tag so people can't get at it by default - it's not something I think has to be checked in, but it does serve as an example of a separate way of integrating DNS serving.

smarterclayton commented:
Also added a multi-Getter resolver that returns the first result identified by a Getter (so that results can be overlaid).

smarterclayton commented:
One more commit - made it so that statistics are optional (so we can integrate with Prometheus) - hope this is ok.

miekg commented on backends/kubernetes/kubernetes.go:
```diff
@@ -0,0 +1,197 @@
+	"github.com/skynetservices/skydns/msg"
+	"github.com/skynetservices/skydns/server"
+
+	"github.com/golang/glog"
```
Any reason to use `glog` here?

smarterclayton replied:
Will remove the example.


> On Mar 8, 2015, at 10:26 AM, Miek Gieben <notifications@github.com> wrote:
> 
> In backends/kubernetes/kubernetes.go:
> 
> > +import (
> > +	"fmt"
> > +	"os"
> > +	"strings"
> > +	"time"
> > +
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
> > +
> > +	"github.com/skynetservices/skydns/msg"
> > +	"github.com/skynetservices/skydns/server"
> > +
> > +	"github.com/golang/glog"
> Any reason to use glog here?
> 
> —
> Reply to this email directly or view it on GitHub.
>

miekg commented on main.go:
```diff
@@ -116,12 +126,25 @@ func main() {
 	}
 }
 
+func loadConfig(client *etcd.Client, config *server.Config) error {
```
I wonder if we should make loadconfig part of the `Getter` interface?

smarterclayton replied:
I tried it but it required gymnastics on initializing the getter (to set priority and ttl once when the getter was created, and then again after the config was loaded).  It felt less coupled this way (I briefly created a second getter but it didn't feel special enough to be independent).

The etcd logic in main could probably move into backends/etcd as helper code though.

> On Mar 8, 2015, at 10:34 AM, Miek Gieben <notifications@github.com> wrote:
> 
> In main.go:
> 
> >  	if err := s.Run(); err != nil {
> >  		log.Fatalf("skydns: %s", err)
> >  	}
> >  }
> >  
> > +func loadConfig(client *etcd.Client, config *server.Config) error {
> I wonder if we should make loadconfig part of the Getter interface?
> 
> —
> Reply to this email directly or view it on GitHub.
>

miekg replied:
[ Quoting <notifications@github.com> in "Re: [skydns] WIP - Abstract record ..." ]
>>  	if err := s.Run(); err != nil {
>>  		log.Fatalf("skydns: %s", err)
>>  	}
>>  }
>>
>> +func loadConfig(client *etcd.Client, config *server.Config) error {
>
>I tried it but it required gymnastics on initializing the getter (to set priority and ttl once when the getter was created, and then again after the config was loaded).  It felt less coupled this way (I briefly created a second getter but it didn't feel special enough to be independent).

Ok. I will try to refactor this later.

>The etcd logic in main could probably move into backends/etcd as helper code though.

Yes, and while doing that I might extend the `Backend` interface a little (or
not).

/Miek

--
Miek Gieben

miekg commented on server/server.go:
```diff
@@ -24,18 +23,59 @@ import (
 }
 
+type Getter interface {
+	Records(name string, exact bool) ([]msg.Service, error)
```
`Getter` sounds so... java-ish, maybe `Backend` or `Backender`?

smarterclayton replied:
Agree, I was only using your example :)

Backend or RecordRetriever or RecordLoader or RecordStore all work for me.


> On Mar 8, 2015, at 10:42 AM, Miek Gieben <notifications@github.com> wrote:
> 
> In server/server.go:
> 
> >  	scache       *cache.Cache
> >  	rcache       *cache.Cache
> >  }
> >  
> > +type Getter interface {
> > +	Records(name string, exact bool) ([]msg.Service, error)
> Getter sounds so... java-ish, maybe Backend or Backender?
> 
> —
> Reply to this email directly or view it on GitHub.
>

miekg replied:
[ Quoting <notifications@github.com> in "Re: [skydns] WIP - Abstract record ..." ]
>>  	scache       *cache.Cache
>>  	rcache       *cache.Cache
>>  }
>>
>> +type Getter interface {
>> +	Records(name string, exact bool) ([]msg.Service, error)
>
>Agree, I was only using your example :)
>
>Backend or RecordRetriever or RecordLoader or RecordStore all work for me.

Ok, let's go with Backend then.

/Miek

--
Miek Gieben

miekg commented:
Do we miss any tests which would have triggered when the reverse lookup was wrong?

Overall I think this looks great and is ready to be merged. I'm not opposed to the kubernetes example, but if we leave it there, we should extend `main.go` to actually make it easy to select. Same for the Getter-layering.

smarterclayton commented:
I think there is a test missing, I'll add it today.

I'll remove Kube (since its not generic enough and can be done outside skydns).

Getter layering is really a tool, someone else could implement it.  It should have tests if it stays, so I'm fine taking it out (fairly obvious code).


> On Mar 8, 2015, at 10:58 AM, Miek Gieben <notifications@github.com> wrote:
> 
> Do we miss any tests which would have triggered when the reverse lookup was wrong?
> 
> Overall I think this looks great and is ready to be merged. I'm not opposed to the kubernetes example, but if we leave it there, we should extend main.go to actually make it easy to select. Same for the Getter-layering.
> 
> —
> Reply to this email directly or view it on GitHub.
>

miekg commented:
I like the Getter layering too. I think it is great to leave it in and add a section about it in the readme.

smarterclayton commented:
Turns out no - the ReverseRecord is needed because the backend may want to handle reverse records in a completely orthogonal way.  The etcd implementation is just simpler, so while the tests function, my custom backend was limited.


> On Mar 8, 2015, at 10:58 AM, Miek Gieben <notifications@github.com> wrote:
> 
> Do we miss any tests which would have triggered when the reverse lookup was wrong?
> 
> Overall I think this looks great and is ready to be merged. I'm not opposed to the kubernetes example, but if we leave it there, we should extend main.go to actually make it easy to select. Same for the Getter-layering.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton commented:
Alright, this is now up to date with the review comments.  I updated the README and removed the Kube section (which was incorrect now).  Let me know if you have any other feedback - you can see how I'm using it here https://github.com/smarterclayton/origin/tree/e04e75f10b18415b7441ba9e12797761167e3905/pkg/dns.  Thanks!

miekg commented:
LGTM

@bketelsen agree?

bketelsen commented:
LGTM

thockin commented:
@smarterclayton Were you or one of your folks going to implement Backend against kube, or are you waiting for me to do it?

smarterclayton commented:
Wanted to talk to you and scope out a plan.  I think the skydns backend is better in general than duplicating etcd records, but the decisions about where it should be located and how it integrates with the cluster are more nuanced.  We can sync on that this week.

The changes here benefit the headless work I started.


> On Apr 6, 2015, at 1:01 AM, Tim Hockin <notifications@github.com> wrote:
> 
> @smarterclayton Were you or one of your folks going to implement Backend against kube, or are you waiting for me to do it?
> 
> —
> Reply to this email directly or view it on GitHub.
>

thockin commented:
Let's chat this week - there are a bunch of DNS-related things in the air
right now.

On Mon, Apr 6, 2015 at 6:24 AM, Clayton Coleman <notifications@github.com>
wrote:

> Wanted to talk to you and scope out a plan. I think the skydns backend is
> better in general than duplicating etcd records, but the decisions about
> where it should be located and how it integrates with the cluster are more
> nuanced. We can sync on that this week.
>
> The changes here benefit the headless work I started.
>
>
> > On Apr 6, 2015, at 1:01 AM, Tim Hockin <notifications@github.com>
> wrote:
> >
> > @smarterclayton Were you or one of your folks going to implement Backend
> against kube, or are you waiting for me to do it?
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/skynetservices/skydns/pull/140#issuecomment-90063532>.
>
